### PR TITLE
IGAPP-1035: Reference open source licenses on web

### PR DESCRIPTION
**Pull request by Cmd190**
Click on Footer>openSourceLicenses to see the license page.

### CODE REVIEW

File: web/src/routes/LicensePage.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  }, [])
+
+  const renderItem = (item: License) => (
+    <LicenseItem name={item.name} license={item.licenses} version={item.version} onPress={item.licenseUrl} />
+  )
+
+  return (
+    <div>
+      <Caption title={t('openSourceLicenses')} />
+      <List items={licenses} renderItem={renderItem} noItemsMessage='TODO' />
```
Comment (Cmd190) [inline web/src/routes/LicensePage.tsx:54]: Should we add an extra translation for "no licenses found" or should we just use an english text without translation? The licenses are english anyway

Comment (f1sh1918) [reply web/src/routes/LicensePage.tsx:55]: i'm fine with english only.

File: web/src/components/LocationFooter.tsx [hunk 2]
```diff
@@ -29,6 +30,7 @@ const LocationFooter: React.FC<PropsType> = ({ city, language, overlay = false }
       <CleanLink to={disclaimerPath}>{t('imprintAndContact')}</CleanLink>
       <CleanLink to={aboutUrl}>{t('settings:about', { appName: buildConfig().appName })}</CleanLink>
       <CleanLink to={privacyUrl}>{t('privacy')}</CleanLink>
+      <CleanLink to={RoutePatterns[LICENSE_ROUTE]}>{t('openSourceLicenses')}</CleanLink>
```
Comment (f1sh1918) [inline web/src/components/LocationFooter.tsx:33]:
```suggestion
      <CleanLink to={RoutePatterns[LICENSE_ROUTE]}>{t('settings:openSourceLicenses')}</CleanLink>
```

File: web/src/routes/LicensePage.tsx (hunk 1 above)
Comment (f1sh1918) [reply web/src/routes/LicensePage.tsx:55]: i think we should keep consistent and add a translation

**Review by f1sh1918 — APPROVED** [review]
deleted

**Review by f1sh1918 — CHANGES_REQUESTED** [review]
Some minor issue with i18n. Rest looks fine

File: .prettierignore [hunk 3]
```diff
@@ -3,6 +3,7 @@ native/ios/**/*
 native/vendor/**/*
 
 native/assets/licenses_native.json
+web/assets/licenses.json
```
Comment (steffenkleinle) [inline .prettierignore:6]: Please be consistent here. Either use `licenses_native.json` and `licenses_web.json` or `licenses.json` in both cases. I'd prefer `licenses.json` since the web/native suffix is really redundant.

File: web/package.json [hunk 4]
```diff
@@ -56,11 +57,14 @@
     "focus-trap-react": "^9.0.2",
     "htmlparser2": "^7.2.0",
     "i18next": "^20.6.1",
+    "license-checker": "^25.0.1",
```
Comment (steffenkleinle) [inline web/package.json:60]: Why do we need this for web but not for native?

File: web/package.json [hunk 5]
```diff
@@ -56,11 +57,14 @@
...
     "lodash": "^4.17.21",
     "mapbox-gl": "npm:empty-npm-package@1.0.0",
     "maplibre-gl": "^2.1.9",
     "moment": "2.29.4",
     "normalize.css": "^8.0.1",
+    "npm-license-crawler": "^0.2.1",
+    "path-to-regexp": "^1.7.0",
```
Comment (steffenkleinle) [inline web/package.json:67]: Why do we need this?

File: web/src/RootSwitcher.tsx [hunk 6]
```diff
@@ -8,6 +8,7 @@ import {
   createCitiesEndpoint,
   JPAL_TRACKING_ROUTE,
   LANDING_ROUTE,
+  LICENSE_ROUTE,
```
Comment (steffenkleinle) [inline web/src/RootSwitcher.tsx:11]: Didn't recognize before but I think this should be `LICENSES_ROUTE` (plural), especially since its value is also `licenses`.

File: web/src/routes/LicensePage.tsx (hunk 1 above)
Comment (steffenkleinle) [reply web/src/routes/LicensePage.tsx:55]: It should never be shown since there will always be more than zero licenses. So I would be fine with just passing an empty string as well.

File: web/src/routes/LicensePage.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  return (
+    <ListItem path={onPress} title={name}>
+      <LicenseContainer>
+        {`version: ${version}`}
+        <br />
+        {`license: ${license}`}
+      </LicenseContainer>
+    </ListItem>
+  )
+}
```
Comment (steffenkleinle) [inline web/src/routes/LicensePage.tsx:35]: Should be moved to its own file or just inlined below.

File: web/src/routes/LicensePage.tsx [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  license: string
+  onPress: string
+}
+
+const LicenseItem = (props: PropsType): ReactElement => {
+  const { name, version, license, onPress } = props
+  return (
+    <ListItem path={onPress} title={name}>
+      <LicenseContainer>
+        {`version: ${version}`}
```
Comment (steffenkleinle) [inline web/src/routes/LicensePage.tsx:29]: Could you try to make this simpler?

File: web/src/routes/LicensePage.tsx [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+type PropsType = {
+  name: string
+  version: string | undefined
+  license: string
+  onPress: string
+}
+
+const LicenseItem = (props: PropsType): ReactElement => {
+  const { name, version, license, onPress } = props
```
Comment (steffenkleinle) [inline web/src/routes/LicensePage.tsx:25]: Optional: If you destructure `props` directly you could remove brackets/return statement.

File: web/src/routes/LicensePage.tsx [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        <br />
+        {`license: ${license}`}
+      </LicenseContainer>
+    </ListItem>
+  )
+}
+
+const LicensePage = (): ReactElement => {
+  const { t } = useTranslation('settings')
+  const [licenses, setLicenses] = useState<License[]>([])
```
Comment (steffenkleinle) [inline web/src/routes/LicensePage.tsx:39]:
```suggestion
  const [licenses, setLicenses] = useState<License[] | null>(null)
```

File: web/src/routes/LicensePage.tsx (hunk 1 above)
Comment (Cmd190) [reply web/src/routes/LicensePage.tsx:55]: good idea

File: web/src/routes/LicensePage.tsx (hunk 9 above)
Comment (Cmd190) [reply web/src/routes/LicensePage.tsx:25]: done

File: web/src/routes/LicensePage.tsx (hunk 10 above)
Comment (Cmd190) [reply web/src/routes/LicensePage.tsx:39]: done

File: web/src/routes/LicensePage.tsx (hunk 8 above)
Comment (Cmd190) [reply web/src/routes/LicensePage.tsx:29]: I tried

File: web/package.json (hunk 5 above)
Comment (Cmd190) [reply web/package.json:67]: I dont't know where this dependencies came from. I removed them and it still works fine

File: .prettierignore (hunk 3 above)
Comment (Cmd190) [reply .prettierignore:6]: done

File: web/src/RootSwitcher.tsx (hunk 6 above)
Comment (Cmd190) [reply web/src/RootSwitcher.tsx:11]: renamed

File: web/src/routes/LicensePage.tsx (hunk 7 above)
Comment (Cmd190) [reply web/src/routes/LicensePage.tsx:35]: Imo it is not necessary to move 20 loc in its own file when it's only used in this file

File: web/src/routes/LicensePage.tsx [hunk 11]
```diff
@@ -21,26 +21,21 @@ type PropsType = {
...
-        {`license: ${license}`}
-      </LicenseContainer>
-    </ListItem>
-  )
-}
+const LicenseItem = ({ license, name, onPress, version }: PropsType): ReactElement => (
+  <ListItem path={onPress} title={name}>
+    <LicenseContainer>
+      <div>{`version: ${version}`}</div>
+      <div>{`license: ${license}`}</div>
```
Comment (steffenkleinle) [inline web/src/routes/LicensePage.tsx:28]: The following should work as well (maybe divs/br tag are needed again):
```suggestion
      version: {version}
      license: ${license}
```

File: web/src/routes/LicensePage.tsx [hunk 12]
```diff
@@ -21,26 +21,21 @@ type PropsType = {
...
+    </LicenseContainer>
+  </ListItem>
+)
 
 const LicensePage = (): ReactElement => {
```
Comment (steffenkleinle) [inline web/src/routes/LicensePage.tsx:33]: Should be the following (same for file name):
```suggestion
const LicensesPage = (): ReactElement => {
```

**Review by steffenkleinle — CHANGES_REQUESTED** [review]
Translation is missing, it should be available at least in German:
image

I think there should be translations for `version` and `license` as well.

I also think you shouldn't use the whole width of the page for the list (probably you just need to wrap it in `<Layout />` or something-

File: web/src/routes/LicensePage.tsx (hunk 11 above)
Comment (Cmd190) [reply web/src/routes/LicensePage.tsx:28]: Well, I think both ways are ok and therefore I leave it as it is

**Comment by Cmd190** [comment]
> Translation is missing, it should be available at least in German: image
> 
> I think there should be translations for `version` and `license` as well.
> 
> I also think you shouldn't use the whole width of the page for the list (probably you just need to wrap it in `<Layout />` or something-



> 

The translation for open source licenses  is available in multiple languages though...
Imo it's not necessary to translate the content of the licenses page. Anyone who want to understand the content must speak english because the packages name aren't translated anyway.

File: native/assets/licenses_native.json [hunk 13]
```diff
@@ -0,0 +1,320 @@
+{
```
Comment (steffenkleinle) [inline native/assets/licenses_native.json:1]: Please remove

File: web/src/routes/LicensePage.tsx (hunk 7 above)
Comment (steffenkleinle) [reply web/src/routes/LicensePage.tsx]: @f1sh1918 @SteffiStuffel what do you think? Do we have/should we have a policy for this?

File: web/src/routes/LicensePage.tsx (hunk 11 above)
Comment (steffenkleinle) [reply web/src/routes/LicensePage.tsx]: Well, please try to keep stuff as simple as possible. It just does not make sense to wrap the whole string in a template string  and use one template in another one if it works otherwise as well.

**Review by steffenkleinle — CHANGES_REQUESTED** [review]
Two more names that just be adjusted:
* `LicenseRouteType` -> `LicensesRouteType` (`api-client/src/routes/index.ts`)
* `LicenseRouteInformationType` -> `LicensesRouteInformationType` (`api-client/src/routes/RouteInformationTypes.ts`)

The page should also show the `GeneralHeader` and the `GeneralFooter`. Sorry, missed that before.

**Comment by steffenkleinle** [comment]
> The translation for open source licenses is available in multiple languages though...

Yes, maybe, but the translations are not used, see my screenshot. 

> Imo it's not necessary to translate the content of the licenses page. Anyone who want to understand the content must speak english because the packages name aren't translated anyway.

Please translate it anyway, we don't use hard coded strings for the UI in our apps for various reasons. We want our apps to be as accessible to all users as possible. That includes that a user can always see and understand what a specific page is about (for example if clicked on a link by mistake) even if the specific content is not understood.

File: web/src/routes/LicensePage.tsx (hunk 11 above)
Comment (Cmd190) [reply web/src/routes/LicensePage.tsx]: I'll just fix it along with the other issues

**Comment by Cmd190** [comment]
> Translation is missing, it should be available at least in German: image
> 
> I think there should be translations for `version` and `license` as well.
> 
> I also think you shouldn't use the whole width of the page for the list (probably you just need to wrap it in `<Layout />` or something-



> 

The translation for open source licenses  is available in multiple languages though...
Imo it's not necessary to translate the content of the licenses page. Anyone who want to understand the content must speak english because the packages name aren't translated anyway. 

> > The translation for open source licenses is available in multiple languages though...
> 
> Yes, maybe, but the translations are not used, see my screenshot. You can fix this by changing
> 
> ```tsx
>  <CleanLink to={RoutePatterns[LICENSES_ROUTE]}>{t('openSourceLicenses')}</CleanLink>
> ```
> 
> to
> 
> ```tsx
>  <CleanLink to={RoutePatterns[LICENSES_ROUTE]}>{t('settings:openSourceLicenses')}</CleanLink>
> ```
> 
> in `GeneralFooter.tsx`.
> 
> > Imo it's not necessary to translate the content of the licenses page. Anyone who want to understand the content must speak english because the packages name aren't translated anyway.
> 
> Please translate it anyway, we don't use hard coded strings for the UI in our apps for various reasons. We want our apps to be as accessible to all users as possible. That includes that a user can always see and understand what a specific page is about (for example if clicked on a link by mistake) even if the specific content is not understood.

ok, I get it now. And hardcoded ui strings are really ugly...

File: web/src/routes/LicensePage.tsx (hunk 7 above)
Comment (Cmd190) [reply web/src/routes/LicensePage.tsx]: I just extracted it into its own file

File: native/package.json [hunk 14]
```diff
@@ -84,7 +84,7 @@
     "react-native-elements": "^3.4.2",
     "react-native-gesture-handler": "^2.5.0",
     "react-native-highlight-words": "^1.0.1",
-    "react-native-inappbrowser-reborn": "^3.7.0",
+    "react-native-inappbrowser-reborn": "^3.6.3",
```
Comment (steffenkleinle) [inline native/package.json:87]: Why the downgrade? Please revert

File: web/src/routes/LicensesPage.tsx [hunk 15]
```diff
@@ -45,11 +26,26 @@ const LicensesPage = (): ReactElement => {
...
+    languageCode,
+    pathname,
+  }
+
   return (
-    <Layout>
-      <Caption title={t('openSourceLicenses')} />
-      <List items={licenses ?? []} renderItem={renderItem} noItemsMessage='' />
-    </Layout>
+    <LocationLayout isLoading={false} {...locationLayoutParams}>
```
Comment (steffenkleinle) [inline web/src/routes/LicensesPage.tsx:45]: This should not be a location layout route since the licenses are not specific to one city/language. This breaks navigation for the link in the general footer e.g. in the city selection:
image

File: api-client/src/routes/pathname.ts [hunk 16]
```diff
@@ -35,7 +35,7 @@ export const cityContentPath = ({ cityCode, languageCode, route, path }: CityCon
   constructPathname([cityCode, languageCode, route, path])
 
 export const pathnameFromRouteInformation = (routeInformation: NonNullableRouteInformationType): string => {
-  if (routeInformation.route === JPAL_TRACKING_ROUTE || routeInformation.route === LICENSES_ROUTE) {
```
Comment (steffenkleinle) [inline api-client/src/routes/pathname.ts]: Revert

**Review by steffenkleinle — CHANGES_REQUESTED** [review]

File: native/package.json [hunk 17]
```diff
@@ -135,6 +135,8 @@
     "jetifier": "^2.0.0",
     "metro-react-native-babel-preset": "^0.66.2",
     "node-fetch": "^3.2.8",
+    "patch-package": "^6.4.7",
+    "postinstall-postinstall": "^2.1.0",
```
Comment (steffenkleinle) [inline native/package.json:139]:
```suggestion
```

File: native/package.json (hunk 14 above)
Comment (f1sh1918) [reply native/package.json]: probably due to merging by accident. reverted

File: api-client/src/routes/RouteInformationTypes.ts [hunk 18]
```diff
@@ -46,7 +46,7 @@ export type NewsRouteInformationType = ParamsType & {
 }
 export type SimpleCityContentFeatureType = ParamsType & {
   // Routes without customizable ids, e.g. '/augsburg/de/disclaimer/
-  route: DisclaimerRouteType | OffersRouteType | SprungbrettOfferRouteType | SearchRouteType
+  route: DisclaimerRouteType | OffersRouteType | SprungbrettOfferRouteType | SearchRouteType | LicensesRouteType
 }
```
Comment (steffenkleinle) [inline api-client/src/routes/RouteInformationTypes.ts:50]: This is not correct, licenses route is not a city content route. Please adjust.

File: translations/translations.json [hunk 19]
```diff
@@ -4427,6 +4427,18 @@
...
   },
+  "licenses": {
+    "de": {
+      "version": "Version:",
+      "license": "Lizenz:",
+      "noLicensesMessage": "Keinen Lizenzinformationen verfügbar"
+    },
+    "en": {
+      "version": "version:",
+      "license": "license:",
```
Comment (steffenkleinle) [inline translations/translations.json:4438]: Since its the first word I'd use uppercase I think
```suggestion
      "version": "Version:",
      "license": "License:",
```

File: web/src/components/KebabMenu.tsx [hunk 20]
```diff
@@ -76,9 +76,13 @@ const ToggleButton = styled.button`
   margin-top: 6px;
 `
 
-const KebabMenu = ({ items, direction }: KebabMenuProps): ReactElement => {
+const KebabMenu = ({ items, direction }: KebabMenuProps): ReactElement | null => {
   const [checked, setChecked] = useState<boolean>(false)
 
+  if (items.length === 0) {
+    return null
+  }
```
Comment (steffenkleinle) [inline web/src/components/KebabMenu.tsx:84]: Why is this change necessary?

File: web/src/components/LocationFooter.tsx [hunk 21]
```diff
@@ -23,12 +23,18 @@ const LocationFooter: React.FC<PropsType> = ({ city, language, overlay = false }
     cityCode: city,
     languageCode: language,
   })
+  const licensesPath = pathnameFromRouteInformation({
+    route: LICENSES_ROUTE,
+    cityCode: city,
+    languageCode: language,
```
Comment (steffenkleinle) [inline web/src/components/LocationFooter.tsx:29]:
```suggestion
```

**Review by steffenkleinle — APPROVED** [review]
Tested :+1:

**Comment by steffenkleinle** [comment]
When I run yarn on this branch the lockfile changes. Apparently it does not in the CI though so probably just a problem on my machine?

**Review by sarahsporck — APPROVED** [review]
Works nice, but currently the licenseUrl doesn't work for repos without a master branch (e.g. dompurify). Opened a [PR](https://github.com/mwittig/license-checker/pull/2).
But i think we can still merge this PR.
Honestly, I'm okay with either approach, but I was wondering why we create the license file beforehand. Was this a performance consideration or do we not now the licenses during runtime?

File: web/src/components/KebabMenu.tsx (hunk 20 above)
Comment (f1sh1918) [reply web/src/components/KebabMenu.tsx:84]: for the `General Header` we do not pass items, so the `KebabMenu` shouldn't be displayed. But maybe we have to remove that soon, because the footer gonna take place in the SidebarMenu.

File: web/src/components/LicenseItem.tsx [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+const LicenseContainer = styled.div`
+  padding: 5px 10px;
+  line-height: 120%;
+`
+
+type PropsType = {
+  name: string
+  version: string | undefined
+  license: string
+  onPress: string
```
[the remark this answers is not in the document]
Comment (f1sh1918) [reply web/src/components/LicenseItem.tsx:16]: Definitely . didn't see this while refactoring ;)

File: api-client/src/routes/RouteInformationTypes.ts (hunk 18 above)
Comment (f1sh1918) [reply api-client/src/routes/RouteInformationTypes.ts:50]: done

**Review by steffenkleinle — APPROVED** [review]

**Review by f1sh1918 — APPROVED** [review]
